### PR TITLE
Enabling LogicalProject pushdown optimizations to eliminate exchange of unused columns

### PR DESCRIPTION
**Pull request by shauryachats**
The current logical planner has a shortcoming for `NOT IN` subqueries where it projects unused columns before the exchange stage, resulting in increased data during shuffles and significantly increasing latency.

An example would be:
```
EXPLAIN PLAN FOR 
SELECT count(*) FROM a 
WHERE a.col1 = 'foo' and a.col2 = 'bar' AND a.col4 = 12 AND a.col5 = false 
AND a.col3 NOT IN (SELECT b.col3 FROM b where b.col3 = 'baz') 
AND a.col3 NOT IN (SELECT b.col3 FROM b where b.col3 = 'qux'))
```
The execution plan is:
```
Execution Plan
PinotLogicalAggregate(group=[{}], agg#0=[COUNT($0)])
  PinotLogicalExchange(distribution=[hash])
    PinotLogicalAggregate(group=[{}], agg#0=[COUNT()])
      LogicalFilter(condition=[IS NOT TRUE($8)])
        LogicalJoin(condition=[=($6, $7)], joinType=[left])
          PinotLogicalExchange(distribution=[hash[6]])
            LogicalProject(col1=[$0], col2=[$1], col4=[$3], col5=[$4], col30=[$5], $f1=[$7], col32=[$2])
              LogicalFilter(condition=[IS NOT TRUE($7)])
                LogicalJoin(condition=[=($5, $6)], joinType=[left])
                  PinotLogicalExchange(distribution=[hash[5]])
                    LogicalProject(col1=[$0], col2=[$1], col3=[$2], col4=[$3], col5=[$4], col30=[$2])
                      LogicalFilter(condition=[AND(=($0, _UTF-8'foo'), =($1, _UTF-8'bar'), =($3, 12), NOT($4))])
                        LogicalTableScan(table=[[default, a]])
                  PinotLogicalExchange(distribution=[hash[0]])
                    PinotLogicalAggregate(group=[{0}], agg#0=[MIN($1)])
                      PinotLogicalExchange(distribution=[hash[0]])
                        PinotLogicalAggregate(group=[{0}], agg#0=[MIN($1)])
                          LogicalProject(col3=[$2], $f1=[true])
                            LogicalFilter(condition=[=($2, CAST(_UTF-8'baz'):INTEGER NOT NULL)])
                              LogicalTableScan(table=[[default, b]])
          PinotLogicalExchange(distribution=[hash[0]])
            PinotLogicalAggregate(group=[{0}], agg#0=[MIN($1)])
              PinotLogicalExchange(distribution=[hash[0]])
                PinotLogicalAggregate(group=[{0}], agg#0=[MIN($1)])
                  LogicalProject(col3=[$2], $f1=[true])
                    LogicalFilter(condition=[=($2, CAST(_UTF-8'qux'):INTEGER NOT NULL)])
                      LogicalTableScan(table=[[default, b]])
```

It can be observed from the execution plan
```
                 PinotLogicalExchange(distribution=[hash[5]])
                    LogicalProject(col1=[$0], col2=[$1], col3=[$2], col4=[$3], col5=[$4], col30=[$2])
```
that while the join is only on `a.col3` and the final result required is a `count(*)`, other columns (`col1`, `col2`, `col4`, `col5`) are present in the `LogicalProject` step right before `PinotLogicalExchange` which should ideally have been pruned.

The cause of the issue is the `LogicalFilter` which does not allow projects to be transposed, since the current rules do not have `PROJECT_FILTER_TRANSPOSE` included. Even if `PROJECT_FILTER_TRANSPOSE` is included in `BASIC_RULES`, it would not work for multiple joins since `BASIC_RULES` optimizations are run only once.

This PR aims to fix this issue by running `Project` pushdown (transpose and merge operations) as a `RuleCollection` after the first iteration of `Filter` pushdown rules are run to ensure `Project` can be pushed down as much as possible. 
After the `Project` pushdown is run, the `Filter` pushdown is run again to push down filters after the project pushdown for two reasons:
- In a generic query, filter pushdowns are generally more selective than project pushdowns and should be given more priority.
- Major changes to the current logical plans should be avoided, and not running a `Filter` pushdown after `Project` pushdown results in a Project pushdown after each logical plan node which increases verbosity.

### CODE REVIEW

**Comment by Jackie-Jiang** [comment]
These rules are repeated in multiple places, and rule sets are executed multiple times, introducing extra overhead and currently very hard to manage. Do you see a better way to re-organize them so that they can cover more general scenarios?

cc @gortiz @yashmayya to also take a look

**Comment by shauryachats** [comment]
@Jackie-Jiang I tried various permutations of reorganizing these rules which could minimize the projection of unused columns and reduce the modification of the number of existing logical plans which this change should not impact.

We have a separate workstream to improve query optimization and we can attempt to tackle it subsequently: https://github.com/apache/pinot/issues/14243

cc: @ankitsultana

**Comment by Jackie-Jiang** [comment]
Basically filter pushdown rules can impact project pushdown rules and vise versa. Should we consider putting them together as one rule collection? Will it potentially cause infinite loop?

**Comment by shauryachats** [comment]
Yes, I tested it out and it does cause an infinite loop since `PROJECT_FILTER_TRANSPOSE` and `FILTER_PROJECT_TRANSPOSE` in the same HepInstruction results in the Project and Filter being transposed one after another infinitely.

File: pinot-query-planner/src/test/resources/queries/JoinPlans.json [hunk 1]
```diff
@@ -294,19 +294,21 @@
           "\n                                LogicalFilter(condition=[=($0, _UTF-8'foo')])",
           "\n                                  LogicalTableScan(table=[[default, b]])",
           "\n              PinotLogicalExchange(distribution=[hash[0]])",
-          "\n                PinotLogicalAggregate(group=[{0}], agg#0=[MIN($1)])",
-          "\n                  PinotLogicalExchange(distribution=[hash[0]])",
-          "\n                    PinotLogicalAggregate(group=[{0}], agg#0=[MIN($1)])",
-          "\n                      LogicalProject(col3=[$2], $f1=[true])",
-          "\n                        LogicalFilter(condition=[=($0, _UTF-8'bar')])",
-          "\n                          LogicalTableScan(table=[[default, b]])",
+          "\n                LogicalProject(col3=[$0], $f1=[$1])",
```
Comment (ankitsultana) [inline pinot-query-planner/src/test/resources/queries/JoinPlans.json:297]: Interesting, I think this project is redundant? Something to note regardless as we work on a proper long term solution. Same in the other parts of this plan.

File: pinot-query-planner/src/test/resources/queries/WindowFunctionPlans.json [hunk 2]
```diff
@@ -3404,7 +3404,7 @@
         "sql": "EXPLAIN PLAN FOR WITH windowfunc AS (SELECT a.col1, ROW_NUMBER() OVER(PARTITION BY a.col2 ORDER BY a.col3) as rownum from a) SELECT a.col1, a.rownum FROM windowfunc AS a where a.rownum < 5",
         "output": [
           "Execution Plan",
-          "\nLogicalProject(col1=[$0], $1=[$3])",
+          "\nLogicalProject(col1=[$0], w0$o0=[$3])",
```
Comment (ankitsultana) [inline pinot-query-planner/src/test/resources/queries/WindowFunctionPlans.json:3407]: @shauryachats : can we also verify that this window query works after this change? (as in it runs successfully)

File: pinot-query-planner/src/test/resources/queries/JoinPlans.json (hunk 1 above)
Comment (shauryachats) [reply pinot-query-planner/src/test/resources/queries/JoinPlans.json:297]: Agreed, we can address this subsequently.

File: pinot-query-planner/src/test/resources/queries/WindowFunctionPlans.json (hunk 2 above)
Comment (shauryachats) [reply pinot-query-planner/src/test/resources/queries/WindowFunctionPlans.json:3407]: Yes, verified that this window query works correctly in Quickstart.

Screenshot 2024-10-22 at 3 56 46 PM

Screenshot 2024-10-22 at 3 56 58 PM

Comment (shauryachats) [reply pinot-query-planner/src/test/resources/queries/WindowFunctionPlans.json:3407]: Yes, verified that window queries work in Quickstart.

Screenshot 2024-10-22 at 3 59 56 PM

Screenshot 2024-10-22 at 4 00 11 PM

**Review by ankitsultana — APPROVED** [review]

**Review by ankitsultana — APPROVED** [review]

**Pull request merged**